src/unit_tests/wazuh_db/test_wdb_upgrade.c

```diff
@@ -63,396 +65,442 @@ int teardown_wdb(void **state) {
-    will_return(__wrap_fopen, 1);
-    will_return(__wrap_fopen, 1);
-    will_return(__wrap_fread, "");
-    will_return(__wrap_fread, 0);
-    expect_any_always(__wrap_fclose, _File);
-    will_return(__wrap_fclose, 0);
-    will_return(__wrap_fclose, 0);
-    expect_any_always(__wrap_chmod, path);
-    will_return(__wrap_chmod, 0);
-    expect_string(__wrap__mwarn, formatted_msg, "Creating Global DB backup and creating empty DB");
-    expect_string(__wrap_unlink, file, "queue/db/global.db");
-    will_return(__wrap_unlink, 0);
-    expect_string(__wrap_wdb_create_global, path, "queue/db/global.db");
-    will_return(__wrap_wdb_create_global, OS_SUCCESS);
-    expect_string(__wrap_sqlite3_open_v2, filename, "queue/db/global.db");
-    expect_value(__wrap_sqlite3_open_v2, flags, SQLITE_OPEN_READWRITE);
-    will_return(__wrap_sqlite3_open_v2, 1);
-    will_return(__wrap_sqlite3_open_v2, SQLITE_OK);
-    expect_string(__wrap_wdb_init, id, "global");
-    will_return(__wrap_wdb_init, (wdb_t*)1);
-    expect_value(__wrap_wdb_pool_append, wdb, (wdb_t*)1);
+    will_return(__wrap_wdb_sql_exec, OS_INVALID);
+    will_return(__wrap_wdb_global_get_most_recent_backup, "test_backup_name");
+    will_return(__wrap_wdb_global_get_most_recent_backup, OS_SUCCESS);
+    expect_string(__wrap_wdb_global_restore_backup, snapshot, "test_backup_name");
+    expect_value(__wrap_wdb_global_restore_backup, save_pre_restore_state, false);
+    will_return(__wrap_wdb_global_restore_backup, OS_SUCCESS);
+    expect_string(__wrap__merror, formatted_msg, "Failed to update global.db to version 2. \
+The global.db was restored to the original state.");
 
     ret = wdb_upgrade_global(data->wdb);
 
-    assert_int_equal(ret, 1);
+    assert_ptr_equal(data->wdb, ret);
+    assert_true(ret->enabled);
 }
 
-void test_wdb_upgrade_global_update_v2_to_v3_success(void **state)
+void test_wdb_upgrade_global_intermediate_upgrade_error(void **state)
 {
     wdb_t *ret = NULL;
     test_struct_t *data  = (test_struct_t *)*state;
 
     expect_string(__wrap_wdb_metadata_table_check, key, "metadata");
-    will_return(__wrap_wdb_metadata_table_check, 2);
+    will_return(__wrap_wdb_metadata_table_check, 1);
 
+    // Getting database version
     expect_string(__wrap_wdb_metadata_get_entry, key, "db_version");
-    will_return(__wrap_wdb_metadata_get_entry, "2");
+    will_return(__wrap_wdb_metadata_get_entry, "1");
     will_return(__wrap_wdb_metadata_get_entry, 1);
+
+    // Creating pre upgrade backup
+    will_return(__wrap_wdb_global_create_backup, "global.db");
+    will_return(__wrap_wdb_global_create_backup, OS_SUCCESS);
+
+    // Error upgrading database
+    expect_string(__wrap__mdebug2, formatted_msg, "Updating database 'global' to version 2");
+    expect_string(__wrap_wdb_sql_exec, sql_exec, schema_global_upgrade_v2_sql);
+    will_return(__wrap_wdb_sql_exec, OS_SUCCESS);
     expect_string(__wrap__mdebug2, formatted_msg, "Updating database 'global' to version 3");
```

Comment:
And add a comment here: **Error upgrading from version 2 to 3**

Reply:
Fixed